File: frontend/src/stores/notification.ts
/**
 * 通知状态管理
 */
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import request from '@/services/request'

export interface Notification {
  id: number
  user_id: number
  type: string
  title: string
  content: string | null
  related_data: Record<string, any> | null
  link: string | null
  is_read: boolean
  created_at: string
  read_at: string | null
}

export interface NotificationPreference {
  id: number
  user_id: number
  email_task_created: boolean
  email_task_updated: boolean
  email_task_assigned: boolean
  email_task_completed: boolean
  email_comment_mentioned: boolean
  email_comment_replied: boolean
  email_project_updated: boolean
  site_task_created: boolean
  site_task_updated: boolean
  site_task_assigned: boolean
  site_task_completed: boolean
  site_comment_mentioned: boolean
  site_comment_replied: boolean
  site_project_updated: boolean
  email_frequency: string
}

export const useNotificationStore = defineStore('notification', () => {
  const notifications = ref<Notification[]>([])
  const unreadCount = ref(0)
  const isLoading = ref(false)
  const preference = ref<NotificationPreference | null>(null)
  
  const hasUnread = computed(() => unreadCount.value > 0)
  
  async function fetchNotifications(unreadOnly = false) {
    isLoading.value = true
    try {
      const params = unreadOnly ? { unread_only: 'true' } : {}
      const res = await request.get('/notifications/', { params })
      notifications.value = res || []
      return res || []
    } catch (error) {
      console.error('获取通知列表失败:', error)
      return []
    } finally {
      isLoading.value = false
    }
  }
  
  async function fetchUnreadCount() {
    try {
      const res = await request.get('/notifications/unread-count')
      unreadCount.value = res.unread_count || 0
      return res.unread_count || 0
    } catch (error) {
      console.error('获取未读数量失败:', error)
      return 0
    }
  }
  
  async function markAsRead(notificationId: number) {
    try {
      await request.post(`/notifications/${notificationId}/read`)
      const notification = notifications.value.find(n => n.id === notificationId)
      if (notification && !notification.is_read) {
        notification.is_read = true
        unreadCount.value = Math.max(0, unreadCount.value - 1)
      }
    } catch (error) {
      console.error('标记已读失败:', error)
    }
  }
  
  async function markAllAsRead() {
    try {
      await request.post('/notifications/read-all')
      notifications.value.forEach(n => {
        n.is_read = true
        n.read_at = new Date().toISOString()
      })
      unreadCount.value = 0
    } catch (error) {
      console.error('标记全部已读失败:', error)
    }
  }
  
  async function deleteNotification(notificationId: number) {
    try {
      await request.delete(`/notifications/${notificationId}`)
      const index = notifications.value.findIndex(n => n.id === notificationId)
      if (index !== -1) {
        if (!notifications.value[index].is_read) {
          unreadCount.value = Math.max(0, unreadCount.value - 1)
        }
        notifications.value.splice(index, 1)
      }
    } catch (error) {
      console.error('删除通知失败:', error)
    }
  }
  
  async function fetchPreferences() {
    try {
      const res = await request.get('/notifications/preferences')
      preference.value = res
      return res
    } catch (error) {
      console.error('获取通知偏好失败:', error)
      return null
    }
  }
  
  async function updatePreferences(prefs: Partial<NotificationPreference>) {
    try {
      const res = await request.put('/notifications/preferences', prefs)
      preference.value = res
      return res
    } catch (error) {
      console.error('更新通知偏好失败:', error)
      return null
    }
  }
  
  function getNotificationIcon(type: string): string {
    const iconMap: Record<string, string> = {
      'task_created': '📋',
      'task_updated': '✏️',
      'task_assigned': '👤',
      'task_completed': '✅',
      'project_created': '📁',
      'project_updated': '📝',
      'comment_mentioned': '💬',
      'comment_replied': '💭'
    }
    return iconMap[type] || '🔔'
  }
  
  function formatTime(dateString: string): string {
    const date = new Date(dateString)
    const now = new Date()
    const diff = now.getTime() - date.getTime()
    
    const minutes = Math.floor(diff / 60000)
    const hours = Math.floor(diff / 3600000)
    const days = Math.floor(diff / 86400000)
    
    if (minutes < 1) return '刚刚'
    if (minutes < 60) return `${minutes}分钟前`
    if (hours < 24) return `${hours}小时前`
    if (days < 7) return `${days}天前`
    
    return date.toLocaleDateString('zh-CN')
  }
  
  return {
    notifications,
    unreadCount,
    isLoading,
    preference,
    hasUnread,
    fetchNotifications,
    fetchUnreadCount,
    markAsRead,
    markAllAsRead,
    deleteNotification,
    fetchPreferences,
    updatePreferences,
    getNotificationIcon,
    formatTime
  }
})

File: backend/app/models/notification.py
"""
通知数据模型
"""
from sqlalchemy import Column, Integer, String, Boolean, DateTime, ForeignKey, Text, JSON
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from app.core.database import Base


class Notification(Base):
    """通知模型"""
    __tablename__ = "notifications"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    
    # 通知类型: task_created, task_updated, task_assigned, task_completed, 
    #           project_created, project_updated, comment_mentioned, comment_replied
    type = Column(String(50), nullable=False, index=True)
    
    # 通知标题和内容
    title = Column(String(200), nullable=False)
    content = Column(Text, nullable=True)
    
    # 关联数据 (JSON格式存储关联的ID和类型)
    # 例如: {"task_id": 1, "project_id": 2}
    related_data = Column(JSON, nullable=True)
    
    # 链接 (点击通知跳转的URL)
    link = Column(String(500), nullable=True)
    
    # 状态
    is_read = Column(Boolean, default=False, index=True)
    is_deleted = Column(Boolean, default=False, index=True)
    
    # 时间戳
    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)
    read_at = Column(DateTime(timezone=True), nullable=True)
    
    # 关联关系
    user = relationship("User", back_populates="notifications")
    
    def __repr__(self):
        return f"<Notification {self.id}: {self.title}>"


class NotificationPreference(Base):
    """用户通知偏好设置"""
    __tablename__ = "notification_preferences"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, unique=True, index=True)
    
    # 邮件通知开关
    email_task_created = Column(Boolean, default=True)
    email_task_updated = Column(Boolean, default=True)
    email_task_assigned = Column(Boolean, default=True)
    email_task_completed = Column(Boolean, default=False)
    email_comment_mentioned = Column(Boolean, default=True)
    email_comment_replied = Column(Boolean, default=True)
    email_project_updated = Column(Boolean, default=True)
    
    # 站内通知开关
    site_task_created = Column(Boolean, default=True)
    site_task_updated = Column(Boolean, default=True)
    site_task_assigned = Column(Boolean, default=True)
    site_task_completed = Column(Boolean, default=True)
    site_comment_mentioned = Column(Boolean, default=True)
    site_comment_replied = Column(Boolean, default=True)
    site_project_updated = Column(Boolean, default=True)
    
    # 通知频率: instant, daily, weekly
    email_frequency = Column(String(20), default="instant")
    
    # 时间戳
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    
    # 关联关系
    user = relationship("User", back_populates="notification_preference")
    
    def __repr__(self):
        return f"<NotificationPreference user_id={self.user_id}>"


class EmailQueue(Base):
    """邮件发送队列"""
    __tablename__ = "email_queue"
    
    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    
    # 邮件类型
    email_type = Column(String(50), nullable=False)
    
    # 收件人信息
    to_email = Column(String(200), nullable=False)
    to_name = Column(String(100), nullable=True)
    
    # 邮件内容
    subject = Column(String(200), nullable=False)
    html_content = Column(Text, nullable=True)
    plain_content = Column(Text, nullable=True)
    
    # 发送状态: pending, sent, failed
    status = Column(String(20), default="pending", index=True)
    
    # 重试次数
    retry_count = Column(Integer, default=0)
    max_retries = Column(Integer, default=3)
    
    # 错误信息
    error_message = Column(Text, nullable=True)
    
    # 时间戳
    created_at = Column(DateTime(timezone=True), server_default=func.now(), index=True)
    sent_at = Column(DateTime(timezone=True), nullable=True)
    
    def __repr__(self):
        return f"<EmailQueue {self.id}: {self.email_type} -> {self.to_email}>"